compiler/rustc_hir_typeck/src/method/prelude_edition_lints.rs

```diff
@@ -35,6 +35,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         let (prelude_or_array_lint, edition) = match segment.ident.name {
             // `try_into` was added to the prelude in Rust 2021.
             sym::try_into if !span.at_least_rust_2021() => (RUST_2021_PRELUDE_COLLISIONS, "2021"),
+            sym::poll
+                if !span.at_least_rust_2024()
+                    && let Some(future) = self.tcx.lang_items().future_trait()
+                    && self
+                        .infcx
+                        .type_implements_trait(future, [self_ty], self.param_env)
+                        .may_apply() =>
```

Comment:
hm, i don't think this is entirely right actually. the self type is `Pin<&mut Self>`, so actually this should probably check for the pin and the mut ref and *then* check whether it implements future?
i'm not entirely sure, what do you think errs?

Reply:
idk, I feel like there are probably quite done codebases that have a fn poll that would get needlessly migrated to worse code, while checking for future here is quite easy

Reply:
That's a very speculative "probably", and "is quite easy" comes with all of the normal side-effects of trait solving when there's infer vars in the receiver like overflows and doing extra work. I'm curious if you have any examples -- most of the `poll`-ish functions that I know have wide adoption are prefixed: `poll_next`, `poll_ready`, etc.

Reply:
I did some grepping of my cargo cache and did find some examples of poll functions.
- https://github.com/rust-lang/regex/blob/8856fe36ac7dc37989e6ffb26b5fc57189bae626/regex-automata/src/util/lazy.rs#L195
- https://github.com/tokio-rs/tokio/blob/c8f3539bc11e57843745c68ee60ca5276248f9f9/tokio/src/runtime/task/harness.rs#L151

I'm just gonna check for `Pin<&mut _>` and avoid the trait solving. That should catch basically all cases.